File: app/src/main/java/net/squanchy/tweets/service/TweetModelConverter.java
package net.squanchy.tweets.service;

import com.twitter.sdk.android.core.models.HashtagEntity;
import com.twitter.sdk.android.core.models.MediaEntity;
import com.twitter.sdk.android.core.models.MentionEntity;
import com.twitter.sdk.android.core.models.Tweet;
import com.twitter.sdk.android.core.models.UrlEntity;

import java.util.ArrayList;
import java.util.List;

import net.squanchy.support.lang.Lists;
import net.squanchy.support.lang.Optional;
import net.squanchy.tweets.domain.TweetLinkInfo;
import net.squanchy.tweets.domain.view.TweetViewModel;
import net.squanchy.tweets.domain.view.User;

public class TweetModelConverter {

    private static final String MEDIA_TYPE_PHOTO = "photo";

    private final TweetSpannedTextBuilder tweetSpannedTextBuilder;

    public TweetModelConverter(TweetSpannedTextBuilder tweetSpannedTextBuilder) {
        this.tweetSpannedTextBuilder = tweetSpannedTextBuilder;
    }

    TweetViewModel toViewModel(Tweet tweet) {
        User user = User.create(tweet.user.name, tweet.user.screenName, tweet.user.profileImageUrlHttps);

        List<Integer> specialCharacters = findSpecialCharacterIndecesFor(tweet.text);
        Range displayTextRange = Range.from(tweet.displayTextRange, tweet.text.length(), specialCharacters.size());
        List<HashtagEntity> hashtags = adjustHashtag(onlyHashtagsInRange(tweet.entities.hashtags, displayTextRange), specialCharacters);
        List<MentionEntity> mentions = adjustMentions(onlyMentionsInRange(tweet.entities.userMentions, displayTextRange), specialCharacters);
        List<UrlEntity> urls = adjustUrls(onlyUrlsInRange(tweet.entities.urls, displayTextRange), specialCharacters);
        List<String> photoUrls = onlyPhotoUrls(tweet.entities.media);
        String displayableText = displayableTextFor(tweet, displayTextRange);

        return TweetViewModel.builder()
                .id(tweet.id)
                .text(displayableText)
                .spannedText(tweetSpannedTextBuilder.applySpans(displayableText, displayTextRange.start(), hashtags, mentions, urls))
                .createdAt(tweet.createdAt)
                .user(user)
                .photoUrl(photoUrlMaybeFrom(photoUrls))
                .linkInfo(TweetLinkInfo.from(tweet))
                .build();
    }

    private List<Integer> findSpecialCharacterIndecesFor(String text) {
        StringBuilder builder = new StringBuilder(text);
        List<Integer> highSurrogateIndices = new ArrayList<>();
        for (int i = 0; i < builder.length() - 1; ++i) {
            if (Character.isHighSurrogate(builder.charAt(i)) && Character.isLowSurrogate(builder.charAt(i + 1))) {
                highSurrogateIndices.add(i);
            }
        }
        return highSurrogateIndices;
    }

    private String displayableTextFor(Tweet tweet, Range displayTextRange) {
        Integer beginIndex = displayTextRange.start();
        Integer endIndex = displayTextRange.end();
        return tweet.text.substring(beginIndex, endIndex);
    }

    private List<HashtagEntity> onlyHashtagsInRange(List<HashtagEntity> entities, Range displayTextRange) {
        return Lists.filter(entities, entity -> displayTextRange.contains(entity.getStart(), entity.getEnd()));
    }

    private List<MentionEntity> onlyMentionsInRange(List<MentionEntity> entities, Range displayTextRange) {
        return Lists.filter(entities, entity -> displayTextRange.contains(entity.getStart(), entity.getEnd()));
    }

    private List<UrlEntity> onlyUrlsInRange(List<UrlEntity> entities, Range displayTextRange) {
        return Lists.filter(entities, entity -> displayTextRange.contains(entity.getStart(), entity.getEnd()));
    }

    private static List<HashtagEntity> adjustHashtag(List<HashtagEntity> entities, List<Integer> indices) {
        List<HashtagEntity> hashtags = new ArrayList<>(entities.size());
        for (HashtagEntity hashtag : entities) {
            int start = hashtag.getStart();
            int offset = 0;
            for (Integer index : indices) {
                if (index - offset <= start) {
                    offset += 1;
                } else {
                    break;
                }
            }
            hashtags.add(new HashtagEntity(hashtag.text, hashtag.getStart() + offset, hashtag.getEnd() + offset));
        }
        return hashtags;
    }

    private static List<MentionEntity> adjustMentions(List<MentionEntity> entities, List<Integer> indices) {
        List<MentionEntity> mentions = new ArrayList<>(entities.size());
        for (MentionEntity mention : entities) {
            int start = mention.getStart();
            int offset = 0;
            for (Integer index : indices) {
                if (index - offset <= start) {
                    offset += 1;
                } else {
                    break;
                }
            }
            mentions.add(new MentionEntity(mention.id, mention.idStr, mention.name, mention.screenName, mention.getStart() + offset, mention.getEnd() + offset));
        }
        return mentions;
    }

    private static List<UrlEntity> adjustUrls(List<UrlEntity> entities, List<Integer> indices) {
        List<UrlEntity> urls = new ArrayList<>(entities.size());
        for (UrlEntity url : entities) {
            int start = url.getStart();
            int offset = 0;
            for (Integer index : indices) {
                if (index - offset <= start) {
                    offset += 1;
                } else {
                    break;
                }
            }
            urls.add(new UrlEntity(url.url, url.expandedUrl, url.displayUrl, url.getStart() + offset, url.getEnd() + offset));
        }
        return urls;
    }

    private List<String> onlyPhotoUrls(List<MediaEntity> media) {
        List<MediaEntity> photos = Lists.filter(media, mediaEntity -> MEDIA_TYPE_PHOTO.equals(mediaEntity.type));
        return Lists.map(photos, mediaEntity -> mediaEntity.mediaUrlHttps);
    }

    private Optional<String> photoUrlMaybeFrom(List<String> urls) {
        if (urls.isEmpty()) {
            return Optional.absent();
        }
        return Optional.of(urls.get(0));
    }

    private static class Range {

        private final int start;
        private final int end;

        static Range from(List<Integer> positions, int textLength, int offset) {
            if (positions.size() != 2) {
                return new Range(0, textLength - 1);
            }
            return new Range(positions.get(0), positions.get(1) + offset);
        }

        private Range(int start, int end) {
            this.start = start;
            this.end = end;
        }

        int start() {
            return start;
        }

        int end() {
            return end;
        }

        boolean contains(int start, int end) {
            return start() <= start && end <= end();
        }
    }
}

File: app/src/main/java/net/squanchy/tweets/service/TweetModelConverter.kt
package net.squanchy.tweets.service

import net.squanchy.service.firebase.model.twitter.FirestoreTweet
import net.squanchy.service.firebase.model.twitter.FirestoreTwitterHashtag
import net.squanchy.service.firebase.model.twitter.FirestoreTwitterMedia
import net.squanchy.service.firebase.model.twitter.FirestoreTwitterMention
import net.squanchy.service.firebase.model.twitter.FirestoreTwitterUrl
import net.squanchy.tweets.domain.TweetLinkInfo
import net.squanchy.tweets.domain.view.TweetViewModel
import net.squanchy.tweets.domain.view.User
import net.squanchy.tweets.view.TweetUrlSpanFactory

private const val MEDIA_TYPE_PHOTO = "photo"

fun mapToViewModel(factory: TweetUrlSpanFactory, tweet: FirestoreTweet): TweetViewModel {
    val user = User(name = tweet.user.name, screenName = tweet.user.screenName, photoUrl = tweet.user.profileImageUrl)

    val emojiIndices = findEmojiIndices(tweet.text)
    val displayTextRange = Range.from(tweet.displayTextRange, tweet.text.length, emojiIndices.size)

    val hashtags = adjustHashtag(onlyHashtagsInRange(tweet.entities.hashtags, displayTextRange), emojiIndices)
    val mentions = adjustMentions(onlyMentionsInRange(tweet.entities.userMentions, displayTextRange), emojiIndices)
    val urls = adjustUrls(onlyUrlsInRange(tweet.entities.urls, displayTextRange), emojiIndices)
    val photoUrls = onlyPhotoUrls(tweet.entities.media)
    val unresolvedPhotoUrl = tweet.entities.media.map { it.url }
    val displayableText = displayableTextFor(tweet.text, displayTextRange, unresolvedPhotoUrl)

    return TweetViewModel(
        id = tweet.id.toLong(),
        text = displayableText,
        spannedText = factory.applySpansToTweet(displayableText, displayTextRange.start(), hashtags, mentions, urls),
        user = user,
        createdAt = tweet.createdAt,
        photoUrl = photoUrlMaybeFrom(photoUrls),
        linkInfo = TweetLinkInfo(tweet)
    )
}

private fun findEmojiIndices(text: String): List<Int> {
    val emojiIndices = mutableListOf<Int>()
    text.forEachIndexed { index, char ->
        if (Character.isHighSurrogate(char) && Character.isLowSurrogate(text[index + 1])) {
            emojiIndices.add(index)
        }
    }
    return emojiIndices
}

private fun onlyHashtagsInRange(entities: List<FirestoreTwitterHashtag>, displayTextRange: Range): List<FirestoreTwitterHashtag> =
    entities.filter { displayTextRange.contains(it.start, it.end) }

private fun onlyMentionsInRange(entities: List<FirestoreTwitterMention>, displayTextRange: Range): List<FirestoreTwitterMention> =
    entities.filter { displayTextRange.contains(it.start, it.end) }

private fun onlyUrlsInRange(entities: List<FirestoreTwitterUrl>, displayTextRange: Range): List<FirestoreTwitterUrl> =
    entities.filter { displayTextRange.contains(it.start, it.end) }

private fun adjustHashtag(entities: List<FirestoreTwitterHashtag>, indices: List<Int>): List<FirestoreTwitterHashtag> {
    for (hashtag in entities) {
        val offset = offsetFrom(hashtag.start, indices)
        hashtag.start = hashtag.start + offset
        hashtag.end = hashtag.end + offset
    }
    return entities
}

private fun adjustMentions(entities: List<FirestoreTwitterMention>, indices: List<Int>): List<FirestoreTwitterMention> {
    for (mention in entities) {
        val offset = offsetFrom(mention.start, indices)
        mention.start = mention.start + offset
        mention.end = mention.end + offset
    }
    return entities
}

private fun adjustUrls(entities: List<FirestoreTwitterUrl>, indices: List<Int>): List<FirestoreTwitterUrl> {
    for (url in entities) {
        val offset = offsetFrom(url.start, indices)
        url.start = url.start + offset
        url.end = url.end + offset
    }
    return entities
}

private fun offsetFrom(start: Int, indices: List<Int>): Int {
    var offset = 0
    indices.takeWhile { it - offset <= start }
        .forEach { offset += 1 }

    return offset
}

private fun onlyPhotoUrls(media: List<FirestoreTwitterMedia>): List<String> {
    return media.filter { it.type == MEDIA_TYPE_PHOTO }
        .map { it.mediaUrl }
}

private fun displayableTextFor(text: String, displayTextRange: Range, photoUrls: List<String>): String {
    val beginIndex = displayTextRange.start()
    val endIndex = displayTextRange.end()
    val displayableText = text.substring(beginIndex, endIndex)
    return removeLastPhotoUrl(displayableText, photoUrls)
}

private fun removeLastPhotoUrl(content: String, photoUrls: List<String>): String {
    if (photoUrls.isEmpty()) {
        return content
    }
    val lastUrl = photoUrls[photoUrls.size - 1]
    return if (content.endsWith(lastUrl)) {
        content.replace(lastUrl, "")
            .trim { it <= ' ' }
    } else {
        content
    }
}

private fun photoUrlMaybeFrom(urls: List<String>): String? =
    if (urls.isEmpty()) null else urls[0]

private class Range(private val start: Int, private val end: Int) {

    internal fun start() = start

    internal fun end() = end

    internal fun contains(start: Int, end: Int): Boolean = start() <= start && end <= end()

    companion object {

        internal fun from(positions: List<Int>, textLength: Int, offset: Int): Range {
            return if (positions.size != 2) {
                Range(0, textLength - 1)
            } else {
                Range(positions[0], positions[1] + offset)
            }
        }
    }
}
